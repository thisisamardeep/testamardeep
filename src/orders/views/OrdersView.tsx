import * as React from 'react';
import {Store} from '../infrastructure/Store';
import {Reducers} from '../store/Reducers';
import {SideEffects} from '../store/SideEffects';
import {CurrencyEnum} from '../enums/CurrencyEnum';
import {Amount} from "./Amount";
import {Currency} from "./Currency";
import {Button} from "./Button";
import {ACTION_AMOUNT_CHANGED, ACTION_BOOKING, ACTION_CURRENCYCHANGED} from "../store/actions";
import {ErrorMessage} from "./ErrorMessage";
import {BookingStatus} from "./BookingStatus";


interface IOrdersViewProps {
}

export interface IOrdersViewState {
    amount: string;
    currencyPair: CurrencyEnum;
    bookingRequest: {
        status: string;
        errorMessage: string;
    };
}

export default class OrdersView extends React.Component<IOrdersViewProps, IOrdersViewState> {
    private store: Store<IOrdersViewState>;

    constructor(props: IOrdersViewProps) {
        super(props);
        const initialState: IOrdersViewState = {
            amount: '10m',
            currencyPair: CurrencyEnum.EURJPY,
            bookingRequest: {
                status: null,
                errorMessage: null
            }
        };
        this.store = new Store(
            initialState,
            Reducers,
            SideEffects,
            (nextState: IOrdersViewState) => this.setState(nextState)
        );
        // set initial state
        this.state = this.store.currentState;
    }

    onAmountChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        this.store.dispatchAction(ACTION_AMOUNT_CHANGED, e.target.value);
    };

    onCurrencyPairChanged = (e: React.ChangeEvent<HTMLSelectElement>) => {
        e.preventDefault();
        this.store.dispatchAction(ACTION_CURRENCYCHANGED, e.target.value);
    };

    onBookRequested = (e: React.MouseEvent<HTMLInputElement>) => {
        e.preventDefault();
        this.store.dispatchAction(ACTION_BOOKING);
    };

    render() {
        const {amount, bookingRequest, currencyPair} = this.state;
        return (
            <div>
                <h1>OrdersList</h1>
                <Amount onAmountChanged={this.onAmountChanged} value={amount}/>
                <Currency currency={currencyPair}
                          onCurrencyPairChanged={this.onCurrencyPairChanged}/>
                <p>Order summary:</p>
                <p>Amount({amount})</p>
                <Button onClick={this.onBookRequested}/>
                <BookingStatus status={bookingRequest.status}/>
                <ErrorMessage errorMessage={bookingRequest.errorMessage}/>
            </div>
        );
    }
}
