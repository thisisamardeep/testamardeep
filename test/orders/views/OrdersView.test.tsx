import React from 'react';
import ReactDOM from 'react-dom';
import {act} from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import OrdersView from '../../../src/orders/views/OrdersView';

let container: any;

beforeEach(() => {
  // @ts-ignore
  window._virtualConsole.emit = jest.fn();
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});


describe('OrdersView', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<OrdersView />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can render, submit and update', () => {
    act(() => {
      ReactDOM.render(<OrdersView />, container);
    });

    const title = container.querySelector('h1');
    const amountLabel = container.querySelector('label:nth-of-type(1)');
    const requestTitle = container.querySelector('h2');

    expect(title.textContent).toBe('OrdersList');
    expect(amountLabel.textContent).toBe('Amount:');
    expect(requestTitle.textContent).toBe('');

    const button = container.querySelector('button');

    // Test second render and componentDidUpdate
    act(() => {
      button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });

    expect(requestTitle.textContent).toBe('Booking In Progress');
  });
});
