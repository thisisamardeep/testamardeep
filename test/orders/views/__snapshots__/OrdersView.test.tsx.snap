// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersView renders correctly 1`] = `
<div>
  <h1>
    OrdersList
  </h1>
  <p>
    <label>
      Amount:
    </label>
    <input
      onChange={[Function]}
      type="text"
      value="10m"
    />
  </p>
  <label>
    Currency:
  </label>
  <select
    onChange={[Function]}
    value="EURJPY"
  >
    <option
      value="EURUSD"
    >
      EURUSD
    </option>
    <option
      value="EURJPY"
    >
      EURJPY
    </option>
    <option
      value="AUDUSD"
    >
      AUDUSD
    </option>
    <option
      value="GBPUSD"
    >
      GBPUSD
    </option>
  </select>
  <br />
  <p>
    Order summary:
  </p>
  <p>
    Amount(
    10m
    )
  </p>
  <p>
    <button
      onClick={[Function]}
    >
      Book
    </button>
  </p>
  <h2 />
  <div />
</div>
`;
